src/types.rs

```diff
@@ -41,6 +43,8 @@ macro_rules! key {
 
 macro_rules! key_with_big_array {
     ($i:ident, $s:expr) => {
+        // Needed to ensure the as_ref_* memory layout is correct.
+        #[repr(transparent)]
```

Comment:
This isn't related to this PR. Can we split into a different one so we don't block the discussion with this?

Also, I'm not sure its the best call to require all types as transparent. When the new PR is created, we should have there the reasoning to do so.

Reply:
This is required for the `as_ref_checked` and `as_ref_unchecked` functions to be correct. See the [docs](https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent). Otherwise you don't know that the memory layout will be correct. The default repr makes no guarantees.

Reply:
@vlopes11 `transparent` is really needed here for correctness. I really don't think it can stop the compiler from making optimizations as this is a unit struct and is always going to have the layout of a `[u8; N]` under the hood. I spoke to @Voxelot about it and he is on board. Does this work for you?

Reply:
Sounds good!